image/postgres/scripts/init-entrypoint.sh

```diff
@@ -2,7 +2,113 @@
+
+    PG_DATA_VERSION=$(cat "${PGDATA}/PG_VERSION")
+
+    if [ "$PG_DATA_VERSION" -lt "$PG_BINARY_VERSION" ]; then
+        # Binaries version is newer, upgrade the data
+
+        # This is the amount of disk space we currently consume. Normally we
+        # could use df as well, since the data will be the only disk space
+        # consumer, but in testing environment it might not be the case.
+        PG_DATA_USED=$(du -s "${PGDATA}" | awk '{print $1}')
+        PG_BACKUP_VOLUME="/backups"
+        echo "Checking backup volume space..."
+
+        # The backup volume needs to accomodate two copies of data, one is the
+        # actual backup, and one is a restored copy, which will be deleted later.
+        if ! check_available_space "${PG_BACKUP_VOLUME}" $((PG_DATA_USED * 2)); then
+            echo "Not enough space. Checking data volume space..."
+            PG_BACKUP_VOLUME="${PGDATA}/../"
+
+            # If no luck, check the main data volume. It has to accomodate two
+            # extra copies of data as well, one is the backup and one is the
+            # restored copy.
+            if ! check_available_space "${PG_BACKUP_VOLUME}" $((PG_DATA_USED * 2)); then
+                echo "Not enough disk space, upgrade is cancelled"
+                exit 1
+            fi
+        fi
+
+        # After this point we know there is enough available disk space.
+
+        BACKUP_DIR="${PG_BACKUP_VOLUME}/backups/$PG_DATA_VERSION-$PG_BINARY_VERSION/"
+        # Do not care about symlinks yet
+        if [ -d "${BACKUP_DIR}" ]; then
+          echo "An upgrade backup directory already exists, skip."
+          else
+            # Do a backup before upgrading. Since the database is stopped we
+            # may as well simple take a filesystem backup. Alternatives would
+            # be pg_dump or pg_basebackup, both require running database
+            # cluster.
+            echo "Backup..."
+            mkdir -p "${BACKUP_DIR}"
+            tar -cf "${BACKUP_DIR}/backup.tar" -C $PGDATA --checkpoint=1000 .
+            sync $BACKUP_DIR/backup.tar
+        fi
+
+        echo "Verify backup..."
+        BACKUP_VERIFY_PGDATA="${BACKUP_DIR}/backup-restore-test"
+        OLD_BINARIES="/usr/lib64/pgsql/postgresql-13/bin/"
+        mkdir -p $BACKUP_VERIFY_PGDATA
+        tar -xvf $BACKUP_DIR/backup.tar -C $BACKUP_VERIFY_PGDATA
+        $OLD_BINARIES/pg_ctl -D $BACKUP_VERIFY_PGDATA -w start
+        $OLD_BINARIES/pg_ctl -D $BACKUP_VERIFY_PGDATA -w stop
+        rm -rf $BACKUP_VERIFY_PGDATA
+
+        # Not sure how it works now, but during the upgrade group permissions
+        # are rejected.
+        chmod 0700 $PGDATA
+
+        echo "Upgrade..."
+        # Good idea to --check first
+        PGSETUP_PGUPGRADE_OPTIONS='-j 4 -k' \
+            PGPASSWORD=$(cat /run/secrets/stackrox.io/secrets/password) \
+            postgresql-upgrade "${PGDATA}"
+
+        # Need to update statistics afterwards
+        # vacuumdb --all --analyze-in-stages
```

Comment:
Should we do a full to clean up space?  Really should we do such a thing before taking the backup so we aren't moving a lot of dead data?

Reply:
Yeah valid point it would make some issues on doing it before.  It would be nearly impossible for someone to manually do it before an operator upgrade.

So would it then make sense to do it after on the new database?  It would mean a longer downtime, but I could see it being beneficial to some larger customers.

Reply:
I was planning to do this indeed after the upgrade, but inside the operator. Any objections to that?

Reply:
Should we also do it in a way that it applies to helm installs?

Reply:
I'm afraid in Helm we would have to confine ourselves to the documentation saying "do that after the upgrade". Any strong opinion about that? Otherwise I'll check who's going to be affected, and if nothing particularly major I would prefer to stick with this approach.

To clarify, I'm not against this in general. But the script has already tons of shell, the more we get the more potential to hit a bug of course. So if possible I would like to keep the upgrade logic to minimum.

Reply:
If it is one command that we just run after the upgrade flow finishes how would it create logic in the init-script?  

Now I could see the argument that the init-script is the wrong place, but I think we could figure out a way to do a vacuum full for every install path and not just the operator.  For instance we could put it in the migrator once it connects to the database.

Reply:
Right, I've expressed myself incorrectly, I was planning to add vacuuming into the migrator (there are some similar bits there already), not in the operator.

Reply:
Perfect that makes a lot more sense.